gladier_xpcs/tools/gather_xpcs_metadata.py

```diff
@@ -6,7 +6,118 @@ def gather_xpcs_metadata(**event):
+
+        def gather_item(name, node):
+            if isinstance(node, h5py.Dataset):
+                key = name.replace('/', '.')
+                if node.shape == ():
+                    try:
+                        items[key] = node[()].decode('utf-8')
+                    except Exception:
+                        try:
+                            items[key] = node[()].item()
+                        except Exception:
+                            items[key] = node[()]
+                if node.shape == (1, 1):
+                    items[key] = decode_dtype(node[0][0], node.dtype)
+                if node.shape in [(1, 2), (1, 3)]:
+                    items[key] = node[0].tolist()
+        items = {}
+        hdf5_dataframe.visititems(gather_item)
+        return items
+
+
+    def get_extra_metadata(metadata):
+        meta = metadata.copy()
+
+        # Add the aps_cycle_v2 key, based on another key below.
+        aps_cycle_key = 'aps_cycle_v2'
+        root_key = 'measurement.instrument.acquisition.root_folder'
+        root = meta[root_key].lstrip('/').rstrip('/')
+        #_, aps_cycle, user_str = root.split('/')
+        if root[-1] == '/':
+            root = root[:-1]
+        aps_cycle, user_str = root.split('/')[-2:]
+        meta[aps_cycle_key] = '/'.join((aps_cycle, user_str))
+
+        # Return the new metadata
+        return meta
+
+
+    def clean_metadata(metadata, spoiled_keys):
+        """Change or delete metadata that meets the following criteria:
+        * Value is NAN --
+            * Reason: Cannot ingest NAN into Globus Search
+            * Result: Change to zero.
+        * Value in SPOILED_KEYS
+            * Reason: Value type has changed in Globus Search since previous ingest
+            * Result: Remove key entirely.
+        """
+        meta = copy.deepcopy(metadata)
+
+        for key in spoiled_keys:
+            if key in meta.keys():
+                meta.pop(key)
+
+        for key, val in meta.items():
+            if any([isinstance(val, t) for t in [int, float]]):
+                if numpy.isnan(val):
+                    meta[key] = 0
+
+        return meta
+
+
+    def gather(dataframe):
+        hframe = h5py.File(dataframe, 'r')
+
+        metafilename, _ = os.path.splitext(os.path.basename(dataframe))
+        metafilename += '.json'
+        metadata = GENERAL_METADATA.copy()
+        metadata.update(gather_items(hframe))
+
+        # Extra stuff we added in later
+        extra_metadata = get_extra_metadata(metadata)
+        metadata.update(extra_metadata)
+        # Keys that cause ingest into Globus Search to fail. This is likely due to
+        # another key of the same name being ingested previously, causing the types
+        # not to match (After first ingest, you cannot ingest a different type).
+        spoiled_keys = ['measurement.instrument.source_begin.datetime']
+        return clean_metadata(metadata, spoiled_keys)
+
```

Comment:
This is a really nice change. For anyone else reading along, this removes the need to install the gladier-xpcs repo on a funcx endpoint since the `from gladier_xpcs.tools.xpcs_metadata import gather` no longer exists. It makes the function large and kinda un-unit-testable, but I think the change is worth it in this case.

Reply:
I agree this is much cleaner.